[ClangImporter][Swiftify] Add _SwiftifyImportProtocol for safe overloads for protocols

hnrklssn opened the pull request:
The existing _SwiftifyImport macro is a peer macro, limiting it to only
    emitting function wrappers in the same scope as the original function.
    Protocols cannot contain function implementations, so these need to be
    placed in a separate protocol extension instead. _SwiftifyImportProtocol
    is an extension macro rather than a peer macro, to enable this
    functionality.
    
    Rather than operating on a single function, like _SwiftifyImport,
    _SwiftifyImportProtocol takes information about multiple methods and
    creates a single protocol extension with all wrappers in a one-shot
    operation. ClangImporter automatically attaches _SwiftifyImportProtocol
    macros to imported protocols with methods annotated with bounds info.
    
    rdar://144335990

hnrklssn committed:
[Swiftify] Add _SwiftifyImportProtocol for safe overloads for protocols

The existing _SwiftifyImport macro is a peer macro, limiting it to only
emitting function wrappers in the same scope as the original function.
Protocols cannot contain function implementations, so these need to be
placed in a separate protocol extension instead. _SwiftifyImportProtocol
is an extension macro rather than a peer macro, to enable this
functionality.

Rather than operating on a single function, like _SwiftifyImport,
_SwiftifyImportProtocol takes information about multiple methods and
creates a single protocol extension with all wrappers in a one-shot
operation.

rdar://144335990

hnrklssn committed:
[ClangImporter] Attach _SwiftifyImportProtocol to imported protocols
with bounds attributes

This creates safe overloads for any methods in the protocol annotated
with bounds information.

rdar://144335990

hnrklssn committed:
[Swiftify] Make safe wrappers explicitly public

hnrklssn commented:
This depends on Clang changes in https://github.com/swiftlang/llvm-project/pull/10068

Xazax-hun commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8865,21 +8869,48 @@ class SwiftifyInfoPrinter {
+      return false;
+
+    printSeparator();
+    out << ".method(name: \"" << Method->getName().getBaseIdentifier()
```
Will this work for overloaded methods?

hnrklssn replied:
Hmm that's a really good point actually... I'm guessing macros don't have access to mangled names of declarations, otherwise comparing the mangled name instead could work.

hnrklssn replied:
This now passes the entire method signature to disambiguate overloaded methods

Xazax-hun commented on lib/Macros/Sources/SwiftMacros/SwiftifyImportMacro.swift:
```diff
@@ -700,370 +700,450 @@ func getParameterIndexForDeclRef(
-        node: expr)
-    }
-    return dotExpr.declName.baseName.text
+func parseEnumName(_ expr: ExprSyntax) throws -> String {
```
Factoring these noop changes (moving methods out to be free functions) out into a separate PR would make reviewing this one with the functional changes much easier.

hnrklssn replied:
These have been merged in #80400 now.

hnrklssn committed:
[Swiftify] Run swift-format (NFC)

hnrklssn committed:
Move registerStdSpanTypeMapping to SwiftifyInfoPrinter (NFC)

`registerStdSpanTypeMapping` used to be a lambda inside `swiftify`.
By moving it, along with the `typeMapping` state, inside
`SwiftifyInfoPrinter` we will simplify changes necessary to support
`_SwiftifyImportProtocol` in ClangImporter.

hnrklssn committed:
Rename .method parameter "name" to "signature"

To disambiguate overloaded methods in _SwiftifyImportProtocol we now
pass the entire function signature instead of just the function name.
This should be reflected in the parameter name.

hnrklssn committed:
Update assert

We previously asserted that a variable contained a FileUnit, but the
intention of this was to separate it from SourceFiles, and SourceFile is
a subtype of FileUnit, so it wouldn't trigger if the variable actually
contained a SourceFile.

hnrklssn committed:
add _SwiftifyProtocolMethodInfo to abi tests

hnrklssn committed:
Pass availability to _SwiftifyImport

This was only passed to _SwiftifyImportProtocol, now it is also
passed to _SwiftifyImport. This results in @available being attached to
generated overloads that use [Mutable][Raw]Span.

hnrklssn committed:
Don't depend on implicit lifetime of Escapable inout

Escapable inout parameters don't have an implicit lifetime that can be
borrowed or copied, because the caller doesn't know what the lifetime of
the new value is, if the callee writes to the parameter. Depending on
the lifetime of an Escapable inout parameter without explicit lifetime
annotation results in an error, so _SwiftifyImport should not do that.

hnrklssn commented:
@swift-ci please smoke test
